com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -104,29 +116,160 @@ public int AddEntry(ulong networkObjectId, int behaviourIndex, int variableIndex
+
+            if (!ret)
+            {
+                //todo: error handling
+            }
+
+            entry.Position = (ushort)pos;
             entry.Length = (ushort)size;
-            FreeMemoryPosition += (int)size;
+        }
+
+        /// <summary>
+        /// Read the buffer part of a snapshot
+        /// Must match WriteBuffer
+        /// The stream is actually a memory stream and we seek to each variable position as we deserialize them
+        /// </summary>
+        /// <param name="reader">The NetworkReader to read our buffer of variables from</param>
+        /// <param name="snapshotStream">The stream to read our buffer of variables from</param>
+        internal void ReadBuffer(NetworkReader reader, Stream snapshotStream)
+        {
+            int snapshotSize = reader.ReadUInt16();
+
+            snapshotStream.Read(Buffer, 0, snapshotSize);
+
+            for (var i = 0; i < LastEntry; i++)
+            {
+                if (Entries[i].Fresh && Entries[i].Key.TickWritten > 0)
+                {
+                    var nv = FindNetworkVar(Entries[i].Key);
+
+                    Stream.Seek(Entries[i].Position, SeekOrigin.Begin);
+
+                    // todo: consider refactoring out in its own function to accomodate
+                    // other ways to (de)serialize
+                    // todo --M1--
+                    // Review whether tick still belong in netvar or in the snapshot table.
+                    nv.ReadDelta(Stream, m_NetworkManager.IsServer);
+                }
+
+                Entries[i].Fresh = false;
+            }
+        }
+
+        /// <summary>
+        /// Read the snapshot index from a buffer
+        /// Stores the entry. Allocates memory if needed. The actual buffer will be read later
+        /// </summary>
+        /// <param name="reader">The reader to read the index from</param>
+        internal void ReadIndex(NetworkReader reader)
+        {
+            Entry entry;
+            short entries = reader.ReadInt16();
+
+            for (var i = 0; i < entries; i++)
+            {
+                entry = ReadEntry(reader);
+                entry.Fresh = true;
+
+                int pos = Find(entry.Key);
+                if (pos == Entry.NotFound)
+                {
+                    pos = AddEntry(entry.Key);
+                }
+
+                // if we need to allocate more memory (the variable grew in size)
+                if (Entries[pos].Length < entry.Length)
+                {
+                    AllocateEntry(ref entry, pos, entry.Length);
+                }
+
+                Entries[pos] = entry;
+            }
+        }
+
+        /// <summary>
+        /// Helper function to find the NetworkVariable object from a key
+        /// This will look into all spawned objects
+        /// </summary>
+        /// <param name="key">The key to search for</param>
+        private INetworkVariable FindNetworkVar(VariableKey key)
```

Comment:
Could this be a static method?  Reason I ask is, this is the only place we really use m_NetworkManager in a significant way, and I would suggest we try and avoid storing a m_NetworkManager ref in each snapshot.

Reply:
I think there's a effect of "Snapshot" being a bit of a misnomer. It's not only instance per tick or anything. There's _only_ one Snapshot instance per connection, plus one instance for the main one.
As we progress, it's not clear yet we'll need one snapshot per connection.
So, having a `m_NetworkManager` reference seems like a pretty small cost. For multi-instance tests, if we put this static, I fear it would come back to haunt us.

Reply:
And in fact, yes, for the branch I'll use to start on the spawning part, there's a single snapshot, the main snapshot to send. Please let me know if it makes this ref more palatable

Reply:
Yeah we have one SnapshotSystem per NetworkManager and not just one for everything so I don't think this should be static.

Reply:
Ok, so then we have one SnapshotSystem per NetworkManager and one Snapshot per SnapshotSystem?  Fine...but then I'm trying to get my arms around the separation of SnapshotSystem and Snapshot.

Reply:
And then, wouldn't this function be more appropriate inside SpawnManager?

Reply:
I would not consider passing `SnapshotSystem` into a `Snapshot`. I'm aiming for nested responsibilities. A `SnapshotSystem` has/knows `Snapshot`, as such, `Snapshot` don't get to know `SnapshotSystem`. Otherwise, This would allow the whole thing to become monolithic. 

I agree that the function could also be inside SpawnManager. Then, it would not take a `VariableKey`, it would take ObjectId + BehaviourIndex + VariableIndex. Because `VariableKey` is a SnapshotSystem concept. Just make the call "Please move to SpawnManager", if you prefer that. :-)